tests/e2e/replication/show_while_creating_invalid_state.py

```diff
@@ -117,7 +118,283 @@ def test_show_replicas(connection):
+            "args": ["--bolt-port", "7688", "--log-level=TRACE"],
+            "log_file": "replica1.log",
+            "setup_queries": ["SET REPLICATION ROLE TO REPLICA WITH PORT 10001;"],
+        },
+        "replica_2": {
+            "args": ["--bolt-port", "7689", "--log-level=TRACE"],
+            "log_file": "replica2.log",
+            "setup_queries": ["SET REPLICATION ROLE TO REPLICA WITH PORT 10002;"],
+        },
+        "replica_3": {
+            "args": ["--bolt-port", "7690", "--log-level=TRACE"],
+            "log_file": "replica3.log",
+            "setup_queries": ["SET REPLICATION ROLE TO REPLICA WITH PORT 10003;"],
+        },
+        "replica_4": {
+            "args": ["--bolt-port", "7691", "--log-level=TRACE"],
+            "log_file": "replica4.log",
+            "setup_queries": ["SET REPLICATION ROLE TO REPLICA WITH PORT 10004;"],
+        },
+        "main": {
+            "args": ["--bolt-port", "7687", "--log-level=TRACE", "--storage-recover-on-startup=true"],
+            "log_file": "main.log",
+            "setup_queries": [],
+            "data_directory": f"{data_directory.name}",
+        },
+    }
+
+    interactive_mg_runner.start_all(CONFIGURATION)
+    cursor = connection(7687, "main").cursor()
+
+    # We want to execute manually and not via the configuration, otherwise re-starting main would also execute these registration.
+    execute_and_fetch_all(cursor, "REGISTER REPLICA replica_1 SYNC TO '127.0.0.1:10001';")
+    execute_and_fetch_all(cursor, "REGISTER REPLICA replica_2 SYNC TO '127.0.0.1:10002';")
+    execute_and_fetch_all(cursor, "REGISTER REPLICA replica_3 ASYNC TO '127.0.0.1:10003';")
+    execute_and_fetch_all(cursor, "REGISTER REPLICA replica_4 ASYNC TO '127.0.0.1:10004';")
+
+    # 1/
+    expected_data = {
+        ("replica_1", "127.0.0.1:10001", "sync", 0, 0, "ready"),
+        ("replica_2", "127.0.0.1:10002", "sync", 0, 0, "ready"),
+        ("replica_3", "127.0.0.1:10003", "async", 0, 0, "ready"),
+        ("replica_4", "127.0.0.1:10004", "async", 0, 0, "ready"),
+    }
+    actual_data = set(execute_and_fetch_all(cursor, "SHOW REPLICAS;"))
+
+    assert actual_data == expected_data
+
+    def check_roles():
+        assert "main" == interactive_mg_runner.MEMGRAPH_INSTANCES["main"].query("SHOW REPLICATION ROLE;")[0][0]
+        for index in range(1, 4):
+            assert (
+                "replica"
+                == interactive_mg_runner.MEMGRAPH_INSTANCES[f"replica_{index}"].query("SHOW REPLICATION ROLE;")[0][0]
+            )
+
+    check_roles()
+
+    # 2/
+    interactive_mg_runner.kill(CONFIGURATION, "main")
+    time.sleep(2)
+
+    # 3/
+    interactive_mg_runner.start(CONFIGURATION, "main")
+    cursor = connection(7687, "main").cursor()
+    check_roles()
+
+    # 4/
+    # We leave some time for the main to recover.
+    time.sleep(2)
+    actual_data = set(execute_and_fetch_all(cursor, "SHOW REPLICAS;"))
+    assert actual_data == expected_data
```

Comment:
replicas may not instantly be in a good state - they might take some time to reconnect and reestablish their configuration. it should be retried a few times in a loop, up to some "definitely-will-never-flake" retry limit, with like 1 second sleeps in between attempts.

Reply:
I have created task T0899-MG to tackle the sleep(). We could replace it overall in the tests but that would go out of scope of this task.